core/src/main/java/com/datastax/oss/driver/api/core/paging/Pager.java

```diff
@@ -0,0 +1,246 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.paging;
+
+import com.datastax.oss.driver.api.core.AsyncPagingIterable;
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class Pager {
+
+  /** A page returned as the result of an offset query. */
+  public interface Page<ElementT> {
+
+    /** The elements in the page. */
+    @NonNull
+    List<ElementT> getElements();
+
+    /**
+     * The page number.
+     *
+     * <p>It will usually match the requested page number, except if it was past the end of the
+     * iterable. In that case, the last page is returned, and the page number will reflect this.
+     */
+    int getPageNumber();
+
+    /** Whether there are other pages after this one. */
+    boolean isLast();
+  }
+
+  /**
+   * Skips the beginning of a set of results to return the given page, assuming the given page size.
+   *
+   * @param iterable the iterable to extract the results from.
+   * @param targetPageNumber the page to return (1 for the first page, 2 for the second page, etc).
+   *     Must be greater than or equal to 1.
+   * @param pageSize the number of elements per page. Must be greater than or equal to 1.
+   * @return the requested page, or the last page if the requested page was past the end of the
+   *     iterable.
+   * @throws IllegalArgumentException if the conditions on the arguments are not respected.
+   */
+  @NonNull
+  public <ElementT> Page<ElementT> getPage(
+      @NonNull PagingIterable<ElementT> iterable, final int targetPageNumber, final int pageSize) {
+
+    throwIfIllegalArguments(iterable, targetPageNumber, pageSize);
+
+    // Holds the contents of the target page. We also need to record the current page as we go,
+    // because our iterable is forward-only and we can't predict when we'll hit the end.
+    List<ElementT> currentPageElements = new ArrayList<>();
+
+    int currentPageNumber = 1;
+    int currentPageSize = 0;
+    for (ElementT element : iterable) {
+      currentPageSize += 1;
+
+      if (currentPageSize > pageSize) {
+        currentPageNumber += 1;
+        currentPageSize = 1;
+        currentPageElements.clear();
+      }
+
+      currentPageElements.add(element);
```

Comment:
This is necessary if the requested page is past the end: we never which actual page we'll end up returning, so we have to keep each page in case it's the last.

Reply:
OK. Then we might be facing one of those rare cases where a `LinkedList` would perform better than `ArrayList`, since `add` is O(1) for the former vs O(n)-ish for the latter.

Reply:
> if the requested page is past the end: we never know which actual page we'll end up returning

Why is that? Shouldn't we just throw an error if the requested page doesn't exist?